pkg/eventpersister/persister/persister-dwh.go

```diff
@@ -0,0 +1,597 @@
+	for {
+		select {
+		case msg, ok := <-p.puller.MessageCh():
+			if !ok {
+				return nil
+			}
+			// dwhReceivedCounter.Inc()
+			id := msg.Attributes["id"]
+			if id == "" {
+				msg.Ack()
+				// TODO: better log format for msg data
+				// dwhHandledCounter.WithLabelValues(codes.MissingID.String()).Inc()
+				continue
+			}
+			if previous, ok := batch[id]; ok {
+				previous.Ack()
+				p.logger.Warn("Message with duplicate id", zap.String("id", id))
+				// dwhHandledCounter.WithLabelValues(codes.DuplicateID.String()).Inc()
+			}
+			batch[id] = msg
+			if len(batch) < p.opts.flushSize {
+				continue
+			}
+			p.send(batch)
+			batch = make(map[string]*puller.Message)
+			timer.Reset(p.opts.flushInterval)
+		case <-timer.C:
+			if len(batch) > 0 {
+				p.send(batch)
+				batch = make(map[string]*puller.Message)
+			}
+			timer.Reset(p.opts.flushInterval)
+		case <-p.ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (p *PersisterDwh) send(messages map[string]*puller.Message) {
+	ctx, cancel := context.WithTimeout(context.Background(), p.opts.flushTimeout)
+	defer cancel()
+
+	envEvents := p.extractEvents(messages)
+	if len(envEvents) == 0 {
+		p.logger.Error("all messages were bad")
+		return
+	}
+
+	evalEvents := []*ecproto.EvaluationEvent{}
+	goalEvents := []*ecproto.GoalEvent{}
+
+	fails := make(map[string]bool, len(envEvents))
+	for environmentNamespace, events := range envEvents {
+		for id, event := range events {
+			switch evt := event.(type) {
+			case *eventproto.EvaluationEvent:
+				e, retriable, err := p.convToEvaluationEvent(ctx, evt, id, environmentNamespace)
+				if err != nil {
+					if err == ErrNoExperiments {
+						p.logger.Warn(
+							"There is no running experiments",
+							zap.Error(err),
+							zap.String("id", id),
+							zap.String("environmentNamespace", environmentNamespace),
+						)
+						continue
+					}
+					if !retriable {
+						p.logger.Error(
+							"failed to convert to evaluation event",
+							zap.Error(err),
+							zap.String("id", id),
+							zap.String("environmentNamespace", environmentNamespace),
+						)
+					}
+					fails[id] = retriable
+					continue
+				}
+				evalEvents = append(evalEvents, e)
+			case *eventproto.GoalEvent:
+				e, retriable, err := p.convToGoalEvent(ctx, evt, id, environmentNamespace)
+				if err != nil {
+					if err == ErrNoExperiments {
+						p.logger.Warn(
+							"There is no running experiments",
+							zap.Error(err),
+							zap.String("id", id),
+							zap.String("environmentNamespace", environmentNamespace),
+						)
+						continue
+					}
+					if !retriable {
+						p.logger.Error(
+							"failed to convert to goal event",
+							zap.Error(err),
+							zap.String("id", id),
+							zap.String("environmentNamespace", environmentNamespace),
+						)
+					}
+					fails[id] = retriable
+					continue
+				}
+				goalEvents = append(goalEvents, e)
+			}
+		}
+	}
+	if err := p.evalEventWriter.AppendRows(ctx, evalEvents); err != nil {
```

Comment:
The function is a bit too long. It's easier to read if you can split the code by checking the event type.

Also, I think we can simplify the conditions because the subscriber subscribes to one event type only. For example, if it subscribes to a goal-event-topic, all the events must be goal events. You don't need to recheck it in the loop.

Please, let me know your thoughts.